### DISCUSSION TITLE
'use client' have an impact on SEO？

### DISCUSSION THREAD

**Discussion by heygsc** [vercel/next.js#67878] [Help] [answered]
### Summary

If we use 'use client' in `all` Server Components of next.js, will SEO become so poor that it is almost the same as React? 

Or will SEO not be affected?

**Comment by icyJoseph** [comment, marked as answer]
No.

'use client' components, are still SSR'd:

- https://github.com/reactwg/server-components/discussions/4

Client Components are the same components we have used up till now. The new thing are Server Components, which exclusively render on a server context.

This also means that you cannot access window, and other DOM APIs, etc. Your code still has to be able to run in both server and browser JavaScript contexts. 

There are techniques to side-step SSR for Client Components though, but it is nothing you get out of the box so to speak:

- https://demystifying-rsc.vercel.app/client-components/next-dynamic/
- https://demystifying-rsc.vercel.app/client-components/no-ssr/

**Reply by icyJoseph** [reply]
No, now it looks like I confused you even more... let's start over.

SSR: Server Side Rendering

Client Components, are the same components we have used until now. In pages routes, components were rendered on the server, either on request-demand or at build time, and then they were also re-rendered client side as part of hydration, and state lifecycles etc.

Because of the server side rendering on the previous step, coding in Next.js, even in Pages Router, required a few guards, for example, one couldn't read from window during render, only in callbacks, interactivity handlers, or effects.

In App Router, Server Components are the new thing, these render on request-demand, or at build time, but not on the client side, and they don't send their JavaScript to the client. 

Client Components, in App Router, are still rendered on request-demand, or at build time, so their HTML is part of what SEO engines would see.

Let's be clear that I am talking about the vanilla case, of course if a client component, needs some kind of user interaction to render, then it won't be part of the initial HTML.

**Reply by heygsc** [reply]
Thank you! I seem to have misunderstood the concept before, I will think about the connections and differences between them again.

**Reply by icyJoseph** [reply]
Great, do not hesitate to ask further, and please do mark as answered if you are happy with the answers so far.

**Reply by little077** [reply]
![微信截图_20240718171738](https://github.com/user-attachments/assets/8f3a0b39-9d33-4e4b-9f2d-a80ea06adcc5)

**Reply by icyJoseph** [reply]
There's a question to consider here, Do you want to make assumptions about SEO/bots/crawlers? Or do you want to facilitate their work?

Some SEO engines, indeed, do run JS, but even those, might have budgets, timeouts, old JS APIs etc, that could prevent it from crawling your page. And this information is often hidden from us.

However, we do know that, if our initial HTML, sent from the server, has everything, then the SEO/bot/crawler engine can easily parse it.

Since SEO is such an important part of business, it does seem to make sense, to want to favor this last bit.

Often when I introduce this to people, I call it, `SEO without assumptions`.

**Reply by little077** [reply]
What you said makes perfect sense

**Reply by heygsc** [reply]
> No, now it looks like I confused you even more... let's start over.
> 
> SSR: Server Side Rendering
> 
> Client Components, are the same components we have used until now. In pages routes, components were rendered on the server, either on request-demand or at build time, and then they were also re-rendered client side as part of hydration, and state lifecycles etc.
> 
> Because of the server side rendering on the previous step, coding in Next.js, even in Pages Router, required a few guards, for example, one couldn't read from window during render, only in callbacks, interactivity handlers, or effects.
> 
> In App Router, Server Components are the new thing, these render on request-demand, or at build time, but not on the client side, and they don't send their JavaScript to the client.
> 
> Client Components, in App Router, are still rendered on request-demand, or at build time, so their HTML is part of what SEO engines would see.
> 
> Let's be clear that I am talking about the vanilla case, of course if a client component, needs some kind of user interaction to render, then it won't be part of the initial HTML.

I noticed that you mentioned in the end that it was under normal circumstances. 
Let's add another limitation. If every 'use client' component has interactions such as useState/onClick, then in this case 'use client' means it won't be used as the initial HTML, so it will have an impact on SEO?

**Reply by little077** [reply]
can you speak chinese?

**Reply by heygsc** [reply]
> can you speak chinese?

哈哈，我没有聊你们俩的话题，我引用的内容还是在沿着原来的内容往下问，他是国外友人，咱们问他问题当然用英文啦。

**Reply by little077** [reply]
ok .Please continue, I am very interested in the SEO section

**Reply by icyJoseph** [reply]
> I noticed that you mentioned in the end that it was under normal circumstances. 
> Let's add another limitation. If every 'use client' component has interactions such as useState/onClick, then in this case 'use client' means it won't be used as the initial HTML, so it will have an impact on SEO?

No. I meant that, if the JSX rendered by a Client Component has branches that activate only when the user does some interaction, or after a browser event, then obviously these JSX branches won't be in the initial HTML.

**Reply by heygsc** [reply]
> > I noticed that you mentioned in the end that it was under normal circumstances.
> > Let's add another limitation. If every 'use client' component has interactions such as useState/onClick, then in this case 'use client' means it won't be used as the initial HTML, so it will have an impact on SEO?
> 
> No. I meant that, if the JSX rendered by a Client Component has branches that activate only when the user does some interaction, or after a browser event, then obviously these JSX branches won't be in the initial HTML.

Oh, I see.

**Reply by heygsc** [reply]
> > I noticed that you mentioned in the end that it was under normal circumstances.
> > Let's add another limitation. If every 'use client' component has interactions such as useState/onClick, then in this case 'use client' means it won't be used as the initial HTML, so it will have an impact on SEO?
> 
> No. I meant that, if the JSX rendered by a Client Component has branches that activate only when the user does some interaction, or after a browser event, then obviously these JSX branches won't be in the initial HTML.

![image](https://github.com/user-attachments/assets/5d70d576-f549-4309-b353-220bac496265)
In this situation where it is not displayed at the beginning and only rendered after clicking on the interaction, the bottom div cannot be seen on the server, right? Does this also mean that the bottom div cannot be seen in SEO crawlers?

**Reply by icyJoseph** [reply]
Right, in the HTML, it won't be present. There would need to be a user action, such as the click, to trigger that.

Though, a bot, or SEO engine, might be instructed to look for a certain element, with a selector, perform an action on it, and crawl that as well. I can't make any assumptions, nor discard this possibility. 

If I were writing a crawler for a given site, and I knew that the info I want, is hidden behind a click event, I'd use a web driver tool, and tell it to click a button, through a selector, and then collect the info I want.

**Reply by heygsc** [reply]
> Right, in the HTML, it won't be present. There would need to be a user action, such as the click, to trigger that.
> 
> Though, a bot, or SEO engine, might be instructed to look for a certain element, with a selector, perform an action on it, and crawl that as well. I can't make any assumptions, nor discard this possibility.
> 
> If I were writing a crawler for a given site, and I knew that the info I want, is hidden behind a click event, I'd use a web driver tool, and tell it to click a button, through a selector, and then collect the info I want.

Wow, thank you. I never thought it would be like this.

**Comment by naingaungphyo** [comment]
I came across this post and I tried to play around with the links @icyJoseph gave in the marked answer comment. But there are still things that I can't confirm. I asked a question in original repo about SSG or ISR not working with `export const revalidate = 90000;`?

https://github.com/matt-kruse/demystifying-rsc/issues/6
